Improvements to conversion reliability

rosetinted opened the pull request:
I encountered a few .caption files that gave problems upon conversion. Some of them used a different type of linebreak, and a more exotic format didn't use double linebreaks to separate subtitles. It could lead to empty or incomplete SRT files being produced.

Most of the conversion code has been rewritten to handle the different formats. It should be a lot more robust now.

Additionally, sometimes not all videos in a course have subtitles. This caused an error message in the console. I added a check for the existence of the .caption file before accessing it.

rosetinted committed:
Fix error message when there is no subtitle to convert.

rosetinted committed:
Fix issue where subtitles with LF linebreaks weren't being converted.

rosetinted committed:
Improvements to conversion reliability

This fixes the following issues with the conversion process, which
could lead to empty, incomplete or incorrect SRT files being produced.

1. Linebreaks other than CRLF are used.
2. Subtitles are separated by single instead of double linebreaks.
3. Linebreak characters can randomly pop up in binary data, at places
   where it becomes too complex to correct them.
4. Instead of the next subtitle, the first non-empty line below a
   subtitle contains a timestamp indicating its end time.
5. The "end time" subtitles are sometimes followed by junk in the text
   area. Often a single null terminator or other nonprintable character.

These issues were solved, or were made easier to solve by no longer
separating subtitles blocks by linebreaks. Instead they are located by
timestamp markers, which has proven more reliable.

mdomnita commented:
Thanks a lot, @rosetinted . 
I finally got to install and test it. It seems to work fine with the decryption of the subtitles and it indeed offers better results.

However, if the /DB option is used (get file and folder names from sqlite database) it only converts the first subtitle file. The others are created with 0b size. Without the db option they turn out fine so it should be a naming/renaming problem. I will look into it and maybe I will figure it out.

Matei

rosetinted commented:
Hm, I've been exclusively using the /DB option. It didn't even occur to me to test without it. If you can't figure it out, let me know what course you're trying to convert and I'll see what happens on my side.

I moved some code around in Decryptor.cs when I added a check for file existence before conversion. Since that part deals with file names, maybe the flaw is there.

rosetinted commented:
@mdomnita 

I ran into the same issue you did while converting more courses. It was caused by timestamps having three digits for the subseconds instead of two, not fitting the usual [##:##:##.##] format and therefore being ignored by the parser. Also the 'fixed' element positions seem to be a little different in these files. I'll see if I can find a workaround.
